crates/wast/src/ast/alias.rs

```diff
@@ -11,50 +11,98 @@ pub struct Alias<'a> {
     /// The kind of item that's being aliased.
-    pub kind: ast::ExportKind,
+    pub kind: AliasKind,
 }
 
-#[derive(Debug)]
+/// aliaskind   ::= (module <id>?)
+///               | (component <id>?)
+///               | (instance <id>?)
+///               | (func <id>?)
+///               | (value <id>?)
+///               | (type <id>?)
+///               | (table <id>?)
+///               | (memory <id>?)
+///               | (global <id>?)
+///               | ... other Post-MVP Core definition kinds
+#[derive(Debug, Clone)]
 #[allow(missing_docs)]
-pub enum AliasSource<'a> {
-    InstanceExport {
-        instance: ast::ItemRef<'a, kw::instance>,
+pub enum AliasKind {
+    Module,
+    Component,
+    Instance,
+    Value,
+    ExportKind(ast::ExportKind)
+}
+
+impl<'a> Parse<'a> for AliasKind {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        let mut l = parser.lookahead1();
+        if l.peek::<ast::ExportKind>() {
+            let kind = parser.parse::<ast::ExportKind>()?;
+            Ok(AliasKind::ExportKind(kind))
+        } else if l.peek::<kw::module>() {
+            parser.parse::<kw::module>()?;
+            Ok(AliasKind::Module)
+        } else if l.peek::<kw::component>() {
+            parser.parse::<kw::component>()?;
+            Ok(AliasKind::Component)
+        } else if l.peek::<kw::instance>() {
+            parser.parse::<kw::instance>()?;
+            Ok(AliasKind::Instance)
+        } else if l.peek::<kw::value>() {
+            parser.parse::<kw::value>()?;
+            Ok(AliasKind::Value)
+        } else {
+            Err(l.error())
+        }
+    }
+}
+
+/// aliastarget ::= export <instanceidx> <name>
+///               | outer <outeridx> <idx>
+#[derive(Debug, Clone)]
+#[allow(missing_docs)]
+pub enum AliasTarget<'a> {
+    Export {
+        instance: ast::Index<'a>,
         export: &'a str,
     },
     Outer {
-        /// The index of the module that this reference is referring to.
-        module: ast::Index<'a>,
-        /// The index of the item within `module` that this alias is referering
-        /// to.
+        /// The number of enclosing components to skip.
+        outer: ast::Index<'a>,
+        /// An index into the target component's `aliaskind` index space.
         index: ast::Index<'a>,
     },
 }
 
 impl<'a> Parse<'a> for Alias<'a> {
     fn parse(parser: Parser<'a>) -> Result<Self> {
-        let span = parser.parse::<kw::alias>()?.0;
-        let source = if parser.parse::<Option<kw::outer>>()?.is_some() {
-            AliasSource::Outer {
-                module: parser.parse()?,
-                index: parser.parse()?,
-            }
-        } else {
-            AliasSource::InstanceExport {
-                instance: parser.parse::<ast::IndexOrRef<_>>()?.0,
-                export: parser.parse()?,
-            }
-        };
-        let (kind, id, name) = parser.parens(|p| Ok((p.parse()?, p.parse()?, p.parse()?)))?;
+        parser.parens(|parser| {
```

Comment:
Conventionally most of the parsers in `wast` delegate surrounding parenthesis parsing to the outer parsing context rather than the inner parsing context, so whomever parses aliases would parse these parens in theory.

Reply:
(this is true for a number of the parsers I see below but I'll just comment here)

I should also point out that this isn't 100% true throughout `wast` but IIRC it's the style I tried to adhere to where possible

Reply:
I personally would like to replace manual paren-consuming and manual token-lookahead with an s-expression parser that produces a tree with strings at the leaves, and then parse that :-).

Reply:
In the meantime would you be ok sticking to to general idioms of the rest of the library as-is?

Reply:
I've now moved most of the parse implementations to this style. The type rules still consume the paren though, because intertype doesn't always start with a paren, and it seems simpler to have it optionally parse a paren than to split the parsing into two places.